platform/linux-generic/arch/aarch64/odp_crypto_armv8.c

```diff
@@ -141,6 +168,182 @@ null_crypto_routine(odp_packet_t pkt ODP_UNUSED,
 	return ODP_CRYPTO_ALG_ERR_NONE;
 }
 
+static
+odp_crypto_alg_err_t aes_gcm_encrypt(odp_packet_t pkt,
+				     const odp_crypto_packet_op_param_t *param,
+				     odp_crypto_generic_session_t *session)
+{
+	armv8_cipher_state_t cs = {
+		.counter = {
+			.d = {0, 0}
+		}
+	};
+	uint8_t *iv_ptr;
+	uint64_t plaintext_bit_length = param->cipher_range.length * 8;
+	uint64_t iv_bit_length = session->p.cipher_iv.length * 8;
+	uint64_t aad_bit_length = session->p.auth_aad_len * 8;
+	uint32_t in_pos = param->cipher_range.offset;
+	uint32_t in_len = param->cipher_range.length;
+	int ret = 0;
+
+	/* Fail early if cipher_range is too large */
+	if (in_len > ARM_CRYPTO_MAX_DATA_LENGTH) {
+		ODP_DBG("ARM Crypto: Packet size too large for requested operation\n");
+		ret = -1;
+		goto err;
+	}
+
+	if (param->cipher_iv_ptr)
+		iv_ptr = param->cipher_iv_ptr;
+	else if (session->p.cipher_iv.data)
+		iv_ptr = session->cipher.iv_data;
+	else
+		return ODP_CRYPTO_ALG_ERR_IV_INVALID;
+
+	cs.constants = &session->cc;
+
+	if (armv8_aes_gcm_set_counter(iv_ptr, iv_bit_length, &cs) != 0) {
+		ODP_DBG("ARM Crypto: Failure while setting nonce\n");
+		ret = -1;
+		goto err;
+	}
+
+	uint32_t seg_len = 0;
+	uint8_t *data = odp_packet_offset(pkt, in_pos, &seg_len, NULL);
+
+	if (in_len > seg_len) {
+		/* Packet is segmented within the cipher range. Copy the cipher
+		 * range to a contiguous buffer. */
+		odp_packet_copy_to_mem(pkt, in_pos, in_len, local.buffer);
+
+		data = local.buffer;
+	}
+
+	if (armv8_enc_aes_gcm_from_state(&cs,
+					 param->aad_ptr, aad_bit_length,
```

Comment:
We have the same (theoretical?) segfault possibility with AAD, at least when it is not a multiple of 16 bytes. A simple way to ensure that things always work would be to copy the AAD into a temporary buffer before giving it to the library. Checking if the address + length is near a page boundary could be another way.

Reply:
Now ARM_CRYPTO_MAX_AAD_LENGTH is 16 even though the maximum AAD length supported is 12. The reason why the macro is 16 and not 12 is not visible anywhere, so someone might later just change it to 12. IOW, you seem to be making an assumption that ARM_CRYPTO_MAX_AAD_LENGTH is divisible by 16 but that assumption is not visible in the code.

Actually, if someone changed ARM_CRYPTO_MAX_AAD_LENGTH to 12, the code would not really break in practice since the temp array is in the stack. There is accessible stack also after the array.

So, I would suggest adding this comment like this before the copying: "Copy AAD in a stack to make sure that the ARM crypto library can read it in 16 byte chunks."

The memset before the copying is unnecessary and I think the actual copying could be done using memcpy.

Reply:
I have made these changes in the next version